src/flash.c

```diff
@@ -5,11 +5,88 @@
  */
 
 #include <flash.h>
+
+#include <log.h>
 #include <target/flash.h>
+#include <private/rom.h>
+
+/* Flash geometry constants */
+#define STUB_FLASH_SECTOR_SIZE          0x1000
+#define STUB_FLASH_BLOCK_SIZE           0x10000
+#define STUB_FLASH_PAGE_SIZE            0x100
+#define STUB_FLASH_STATUS_MASK          0xFFFF
+
+/* ROM */
+
+extern int esp_rom_spiflash_config_param(uint32_t device_id,
+                                         uint32_t chip_size, uint32_t block_size, uint32_t sector_size,
+                                         uint32_t page_size, uint32_t status_mask);
+
+static uint32_t device_id_to_flash_size(uint32_t device_id)
```

Comment:
How about this version? Less text and data usage. Since this is a stub code and size always metters, we should try to use memory like we have nothing :)

```
call `device_id_to_flash_size(device_id && 0xff);` 
static uint32_t device_id_to_flash_size(uint32_t device_id)
{
    STUB_LOG_TRACEF("device_id: 0x%x\n", device_id);

    switch (device_id) {
    case 0x12:
        return 256 * 1024;
    case 0x13:
        return 512 * 1024;
    case 0x14:
    case 0x15:
    case 0x16:
    case 0x17:
    case 0x18:
    case 0x19:
    case 0x20:
    case 0x21:
    case 0x22:
        return (device_id + 1 - 0x14) * 1024 * 1024;
    case 0x39:
        return 32 * 1024 * 1024;
    }

    STUB_LOGE("Unknown device_id! %x", device_id);

    return 0;
}
```

Reply:
sure, good point

Reply:
Only thing, let me put `device_id && 0xff` inside device_id_to_flash_size. Otherwise I stumble when looking at the code. If no other reason